ethportal-peertest/src/lib.rs

```diff
@@ -2,31 +2,75 @@ pub mod cli;
 pub mod jsonrpc;
 
 pub use cli::PeertestConfig;
-pub use jsonrpc::{get_enode, test_jsonrpc_endpoints_over_http, test_jsonrpc_endpoints_over_ipc};
+pub use jsonrpc::get_enode;
 
 use std::sync::Arc;
 
+use futures::future;
+
 use trin_core::cli::TrinConfig;
 use trin_core::jsonrpc::service::JsonRpcExiter;
 
-pub async fn launch_buddy_node() -> (String, Arc<JsonRpcExiter>) {
+pub struct PeertestNode {
+    pub enr: String,
+    pub exiter: Arc<JsonRpcExiter>,
+}
+
+pub struct AllPeertestNodes {
+    pub bootnode: PeertestNode,
+    pub nodes: Vec<PeertestNode>,
+}
```

Comment:
The name `AllPeertestNodes` sort of implies that all the nodes in the struct are homogeneous, and it does not convey that the boot node is distinct. I think it would be more clear if we used a different name, or if we removed the struct and handled the boot node separately from the other nodes.

Reply:
Any suggestions for a different name? `AllTestnetNodes` (but that's pretty much the same name). I threw them together in a single struct, since it's easier to pass around a single struct rather than a bootnode and the other nodes separately. Imo, there really is not a big difference between a bootnode and a regular node, they are mostly homogeneous.

Reply:
I think we should remove the struct altogether. It doesn't seem so useful.

The function `validate_endpoint_response` does not even use the argument.

The function `all_endpoints` only accesses the `bootnode` field of the argument.

If we need the two in a single type, like the return type for `launch_peertest_nodes`, then we can use a tuple struct.

Reply:
Fair enough, I was building with patterns in mind whose application were beyond this actual pr. So while they weren't immediately useful, I was building around the idea that they will be useful soon. Well, I've updated the general structure of our tests to exhibit why something like `AllTestnetNodes` is actually useful (imo).

The main problem with our testing architecture is that it was setup in such a way that you can only execute a single test for each jsonrpc endpoint. This is obviously not sustainable, as we will want to support multiple tests for a single endpoint. Case in point, we now test `portal_*FindNodes` with distances (`[0]` -> returns the peer nodes enr & `[256]` -> returns enrs from the peer nodes routing table). 

So another round of :eyes: on the new testing structure would be helpful, there might be a simpler way to set this up, but the blocker I ran into was that the compiler treats every closure as an independent type, so the only way to have them all pass the compiler check as similar types is to wrap the validation logic into a function. @carver If you have any thoughts on the new setup, I'd also be curious to hear what you think since you've spent some time in this part of the codebase.

Also, once we start including `FindContent`/`Content` tests, I believe this struct will also prove useful for storing test data. Now, maybe the problem is with the name? Something like `PeertestConfigData` might be more appropriate?

> If we need the two in a single type, like the return type for launch_peertest_nodes, then we can use a tuple struct.

I always prefer named structs over unnamed tuples. It's minimal overhead, it's easier to read, and you don't have to keep track of what type is in which index. For me, a rule of thumb, is avoid destructuring a tuple with indexes at all costs... 
```rust
let things = Things{ one_thing, another_thing};
...
do_something(things.one_thing);
```
is much better than
```rust
let things = (one_thing, another_thing);
...
do_something(things[0])
```

Reply:
I didn't mean to imply that that sort of structure would never be useful, but a change to the testing structure belongs in a separate PR. The change in structure looks beneficial, but it's harder to evaluate when those changes are not isolated. The focus here is just on `FindNodes`.

I did not like the name, but that was a minor issue. The main issue was that the struct did not provide much utility. With this new structure, it seems like the struct would be useful, but like I said above, I think we should address that in a separate PR.

For the proposed use of the tuple struct, where we return a tuple from the function, we would not lose clarity in the way that your example describes. It would look like the following:

```rust
let (bootnode, other_nodes) = launch_peertest_nodes(2);
```